content/2016.09.trying-kinto.md

```diff
@@ -0,0 +1,420 @@
+$ cd kinto
+$ sed -i s,debian:sid/armhf,debian:sid, Dockerfile
+$ docker build -t croco/kinto:latest .
+```
+Je l'ai ensuite lancé avec le fichier d'environnement suivant:
+```
+$ docker run --env-file kinto.env --link kinto_db:db -p 127.0.0.1:8888:8888 --name kinto_web croco/kinto:latest
+```
+
+```
+$ cat kinto.env
+KINTO_CACHE_BACKEND=kinto.core.cache.postgresql
+KINTO_CACHE_URL=postgres://postgres:postgres@db/postgres
+KINTO_STORAGE_BACKEND=kinto.core.storage.postgresql
+KINTO_STORAGE_URL=postgres://postgres:postgres@db/postgres
+KINTO_PERMISSION_BACKEND=kinto.core.permission.postgresql
+KINTO_PERMISSION_URL=postgres://postgres:postgres@db/postgres
+KINTO_USERID_HMAC_SECRET=SECRET
+KINTO_BATCH_MAX_REQUESTS=200
+```
+
+Le serveur kinto n'est accessible qu'en local (`-p 127.0.0.1:8888:8888`). Un serveur nginx sert de reverse-proxy,
+de point de terminaison SSL et bloque certaines URL.
+Je me suis basé sur les
+[recommandations](http://kinto.readthedocs.io/en/stable/configuration/production.html#using-nginx) de la doc, mais je
+n'ai pas réussi à configurer uwsgi (et j'ai cassé virtualenv en mélangeant du python2 et 3).
+
+La config nginx :
+```
+upstream kinto {
+     server 127.0.0.1:8888;
+}
+
+server {
+	listen 80;
+
+	server_name kinto.elelay.fr;
+	root /var/www/kinto;
+
+	return 302 https://kinto.elelay.fr$request_uri;
+}
+
+server {
+	listen 443 ssl;
+
+	server_name kinto.elelay.fr;
+	ssl_certificate /etc/letsencrypt/live/trackers.elelay.fr/fullchain.pem;
+	ssl_certificate_key /etc/letsencrypt/live/trackers.elelay.fr/privkey.pem;
+	root /var/www/kinto;
+
+	charset utf-8;
+
+	client_max_body_size 75M;
+	
+	location /.well-known {
+		allow all;
+	}
+	
+	location / {
+		proxy_pass http://kinto;
+	}
+
+	location = / {
+		deny all;
+	}
+
+	location ~ /v1/__(.+)__ {
+		allow 127.0.0.1;
+		deny all;
+	}
+}
+```
+
+# Client kinto.js
+
+Mon objectif était d'abord d'en apprendre le maximum sur kinto.js. J'ai bien aimé le coté rustique du tutoriel (pas de
+dépendance à n frameworks) et suis resté sur cette simplicité :
+
+ - le CSS de bootstrap, les icônes de fontawesome
+ - quelques utilitaires : moment, lodash,
+ - datatables (après avoir fait ma propre gestion de table, mais voulant déplier certaines lignes de la table et ayant trouvé
+   un exemple tout prêt pour ce faire).
+
+![Capture d'écran de l'application]({filename}/images/trying-kinto-trackers-app.png)
+
+
+## Etat initial
+
+Lorsque vous accédez pour la première fois à l'[application](https://trackers.elelay.fr/trackers.html), un panneau vous
+est présenté, proposant de récupérer les tickets dont éventuellement les vieux tickets.
+
+Petite astuce pour déterminer si c'est la première fois :
+```javascript
+tickets.db.getLastModified().then(function(res) {
+	if (res) {
+		hasDB();
+	} else {
+		noDB();
+	}
+});
+```
+
+Pour afficher le nombre de tickets sur le serveur avant de les récupérer, il faut descendre dans le package
+[kinto-http](https://github.com/Kinto/kinto-http.js).
```

Comment:
Super: ça simplifie sacrément la récupération du Total-Records initial :
```
var api = tickets.api.bucket(tickets.bucket).collection(tickets.name);
var total = api.getTotalRecords();
```
Pour obtenir le nb de documents mis à jour (ajouter le since), on fait comment ? Le paramètre options n'aide pas dans ce cas (on peut ajouter `headers: { 'If-None-Match': lastModified}` pour ne pas retourner le nb si pas modifié mais on n'a pas la possibilité de compter le nb de docs modifiés depuis `lastModified`).
si #127 est implémenté en passant tous les params à endpoint dans un objet, il pourrait tester la présence d'un paramètre `since`...
Sinon, tester simplement la présence de since dans `options` de [getTotalRecords](https://github.com/Kinto/kinto-http.js/blob/v2.5.0/src/collection.js#L88).